### ISSUE TITLE
Data imbalance

### ISSUE THREAD

**Issue by Jestx170** [kaopanboonyuen/CHULA#1]
มีวิธีไหนบ้างนะครับถ้าไม่ใช้เพิ่มโมเดลอีกอัน

**Comment by kaopanboonyuen** [comment]
**Hey @Jestx170**

I'll break them down below.

---

### 1. **Class Weighting (Loss Function Modification)**

You can modify the loss function to assign more importance to underrepresented classes by using class weights. This makes the model focus more on the minority classes during training.

**Example code:**

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')

# Define class weights (example: 1.0 for class 0, 5.0 for class 1, etc.)
class_weights = [1.0, 5.0, 3.0, 1.0]

model.train(data='data.yaml', batch=16, epochs=50, weights='yolov8n.pt', hyp={'cls_loss': class_weights})
```

---

### 2. **Oversampling Minority Classes**

If certain classes are underrepresented, we can duplicate their images during training. This way, the model sees more of those minority classes and can better learn their features.

**Example code:**

```python
from ultralytics import YOLO
import random

# Function to oversample minority class
def oversample_dataset(dataset):
    minority_class = 1
    minority_images = [img for img in dataset if img['label'] == minority_class]
    augmented_images = random.sample(minority_images, len(minority_images))
    dataset.extend(augmented_images)
    return dataset

model = YOLO('yolov8n.pt')
dataset = model.load_data('path_to_data')
augmented_data = oversample_dataset(dataset)
model.train(data=augmented_data, epochs=50)
```

---

### 3. **Undersampling Majority Classes**

You can also reduce the number of samples from overrepresented classes to make the dataset more balanced. This can help prevent the model from becoming biased toward the majority class.

**Example code:**

```python
from ultralytics import YOLO
import random

def undersample_dataset(dataset):
    majority_class = 0
    majority_images = [img for img in dataset if img['label'] == majority_class]
    undersampled_images = random.sample(majority_images, len(majority_images) // 2)
    dataset = [img for img in dataset if img['label'] != majority_class] + undersampled_images
    return dataset

model = YOLO('yolov8n.pt')
dataset = model.load_data('path_to_data')
undersampled_data = undersample_dataset(dataset)
model.train(data=undersampled_data, epochs=50)
```

---

### 4. **Data Augmentation**

Data augmentation can generate new images from existing data by applying random transformations like flipping, cropping, or changing brightness/contrast. This works well for minority classes, as it increases their variety.

**Example code:**

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')

# Enabling augmentations
model.train(data='data.yaml', epochs=50, augment=True)
```

---

### 5. **Focal Loss**

Focal Loss helps the model focus on difficult-to-classify examples, especially those that are misclassified. It's great for imbalanced datasets because it reduces the weight of well-classified examples.

**Example code:**

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')

# Setting up Focal Loss parameters (you may need to adjust based on your needs)
focal_loss = {
    'cls_loss': 1.0,
    'fl_loss': 0.25,
    'gamma': 2.0
}

model.train(data='data.yaml', epochs=50, hyp=focal_loss)
```

---

### 6. **Hard Negative Mining**

This technique involves selecting difficult-to-classify "negative" samples (examples the model thinks are negative but are hard to classify) and feeding them to the model. This helps the model learn to detect harder cases.

**Example code:**

```python
from ultralytics import YOLO

def hard_negative_mining(dataset):
    hard_negatives = [img for img in dataset if img['predictions']['confidence'] < 0.1]
    dataset.extend(hard_negatives)
    return dataset

model = YOLO('yolov8n.pt')
dataset = model.load_data('path_to_data')
dataset_with_hard_negatives = hard_negative_mining(dataset)
model.train(data=dataset_with_hard_negatives, epochs=50)
```

---

### 7. **SMOTE (Synthetic Minority Over-sampling Technique)**

SMOTE creates synthetic samples for the minority class by interpolating between existing examples. It's a good way to balance the dataset without just duplicating images.

**Example code:**

```python
from imblearn.over_sampling import SMOTE
from ultralytics import YOLO
import numpy as np

X_train, y_train = load_data('data.yaml')

smote = SMOTE(sampling_strategy='auto', random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_train, y_train)

model = YOLO('yolov8n.pt')
model.train(X=X_resampled, y=y_resampled, epochs=50)
```

---

### 8. **Transfer Learning**

Pre-train the model on a large, balanced dataset (e.g., COCO, ImageNet), and then fine-tune it on your imbalanced dataset. The model already has a general understanding of object features, so it’s easier for it to adapt to your specific dataset.

**Example code:**

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')

# Fine-tune on your imbalanced dataset
model.train(data='data.yaml', epochs=50, transfer=True)
```

---

PS:
This is my repo for my new research paper—feel free to explore! All of this is part of the SC310005_ArtificialIntelligence_2025s1 class. For future discussions, please post in the course GitHub repo.

---
